homeassistant/components/esphome/__init__.py

```diff
@@ -199,93 +200,238 @@ async def on_login() -> None:
+        async with self._tries_lock:
+            tries = self._tries
+            self._tries += 1
 
         try:
-            await cli.connect(on_stop=try_connect, login=True)
+            await self._cli.connect(on_stop=self._on_disconnect, login=True)
         except APIConnectionError as error:
             level = logging.WARNING if tries == 0 else logging.DEBUG
             _LOGGER.log(
                 level,
                 "Can't connect to ESPHome API for %s (%s): %s",
-                entry.unique_id,
-                host,
+                self._entry.unique_id,
+                self._host,
                 error,
             )
             # Schedule re-connect in event loop in order not to delay HA
             # startup. First connect is scheduled in tracked tasks.
-            data.reconnect_task = hass.loop.create_task(
-                try_connect(tries + 1, is_disconnect=False)
-            )
+            async with self._wait_task_lock:
+                # Allow only one wait task at a time
+                # can happen if mDNS record received while waiting, then use existing wait task
+                if self._wait_task is not None:
+                    return
+
+                self._wait_task = self._hass.loop.create_task(
+                    self._wait_and_start_reconnect()
+                )
         else:
-            _LOGGER.info("Successfully connected to %s", host)
-            hass.async_create_task(on_login())
+            _LOGGER.info("Successfully connected to %s", self._host)
+            async with self._tries_lock:
+                self._tries = 0
+            async with self._connected_lock:
+                self._connected = True
+            self._hass.async_create_task(self._on_login())
+
+    async def _reconnect_once(self):
+        # Wait and clear reconnection event
+        await self._reconnect_event.wait()
+        self._reconnect_event.clear()
+
+        # If in connected state, do not try to connect again.
+        async with self._connected_lock:
+            if self._connected:
+                return False
+
+        # Check if the entry got removed or disabled, in which case we shouldn't reconnect
+        if self._entry.entry_id not in self._hass.data[DOMAIN]:
+            # When removing/disconnecting manually
+            return
+
+        device_registry = self._hass.helpers.device_registry.async_get(self._hass)
+        devices = dr.async_entries_for_config_entry(
+            device_registry, self._entry.entry_id
+        )
+        for device in devices:
+            # There is only one device in ESPHome
+            if device.disabled:
+                # Don't attempt to connect if it's disabled
+                return
+
+        await self._try_connect()
+
+    async def _reconnect_loop(self):
+        while True:
+            try:
+                await self._reconnect_once()
+            except asyncio.CancelledError:  # pylint: disable=try-except-raise
```

Comment:
From Python 3.8 CancelledError inherits BaseException instead of Exception. So we don't need this case.

Reply:
This is not to catch CancelledExceptions, but rather so that if the reconnect task is cancelled (meaning the reconnect loop should stop) we will actually cancel the loop, not just log the error and continue forever.

Reply:
Yes, `except Exception:` won't catch `asyncio.CancelledError`.

Reply:
Ah sorry, yes now I understand.

Thanks, didn't know that! Will fix this later.